### PR TITLE
enhance: segment alloc interceptor

### PR DESCRIPTION
**Pull request by chyezh**
#33285

- add segment alloc interceptor for streamingnode.
- add add manual alloc segment rpc for datacoord.

### CODE REVIEW

**Comment by chyezh** [comment]
related PR: #34998 35002

File: internal/streamingnode/server/wal/interceptors/segment/manager/seal_queue.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	if err != nil {
+		return errors.Wrap(err, "at create new flush message")
+	}
+
+	msgID, err := m.wal.Get().Append(ctx, msg)
+	if err != nil {
+		m.logger.Warn("send flush message into wal failed", zap.Int64("collectionID", collectionID), zap.String("vchannel", vchannel), zap.Int64s("segmentIDs", segmentIDs), zap.Error(err))
+		return err
+	}
+	m.logger.Info("send flush message into wal", zap.Int64("collectionID", collectionID), zap.String("vchannel", vchannel), zap.Int64s("segmentIDs", segmentIDs), zap.Any("msgID", msgID))
```
Comment (jaime0815) [inline internal/streamingnode/server/wal/interceptors/segment/manager/seal_queue.go:177]: Use debug level instead Info level

File: internal/streamingnode/server/wal/interceptors/segment/policy/segment_limitation_policy.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+// jiiterSegmentLimitationPolicy is the policy to generate the limitation of the segment.
+// Add a jitter to the segment size limitation to scatter the segment sealing time.
+type jitterSegmentLimitationPolicy struct{}
+
+// GenerateLimitation generates the limitation of the segment.
+func (p jitterSegmentLimitationPolicy) GenerateLimitation() SegmentLimitation {
+	// TODO: It's weird to set such a parameter into datacoord configuration.
+	// Refactor it in the future
+	jitter := paramtable.Get().DataCoordCfg.SegmentSealProportionJitter.GetAsFloat()
+	jitterRatio := 1 - jitter*rand.Float64() // generate a random number in [1-jitter, 1]
```
Comment (jaime0815) [inline internal/streamingnode/server/wal/interceptors/segment/policy/segment_limitation_policy.go:44]: default jitter is 0.1， so jitterRatio may be too large.  is it better to jitterRatio = [jitter - rand[0.1-0.3], jitter+  rand[0.1-0.3]] ?

**Comment by jaime0815** [comment]
/lgtm

Resolve outstanding issues in the next PR.

File: internal/streamingnode/server/wal/interceptors/segment/manager/partition_manager.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	shouldBeSealedSegments := make([]*segmentAllocManager, 0, len(m.segments))
+	segments := make([]*segmentAllocManager, 0, len(m.segments))
+	for _, segment := range m.segments {
+		// A already sealed segment may be came from recovery.
+		if segment.GetState() == streamingpb.SegmentAssignmentState_SEGMENT_ASSIGNMENT_STATE_SEALED {
+			shouldBeSealedSegments = append(shouldBeSealedSegments, segment)
+			m.logger.Info("segment has been sealed, remove it from assignment",
```
[the remark this answers is not in the document]
Comment (chyezh) [reply internal/streamingnode/server/wal/interceptors/segment/manager/partition_manager.go:77]: it's a segment level log, so there will never generate too much logs.
every seal operations will not contain much segments by design.

File: internal/streamingnode/server/wal/interceptors/segment/manager/partition_manager.go [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+	m.segments = make([]*segmentAllocManager, 0)
+	return canBeSealed
+}
+
+// hitSealPolicy checks if the segment should be sealed by policy.
+func (m *partitionSegmentManager) hitSealPolicy(segmentMeta *segmentAllocManager) bool {
+	stat := segmentMeta.GetStat()
+	for _, p := range policy.GetSegmentAsyncSealPolicy() {
+		if result := p.ShouldBeSealed(stat); result.ShouldBeSealed {
+			m.logger.Info("segment should be sealed by policy",
```
[the remark this answers is not in the document]
Comment (chyezh) [reply internal/streamingnode/server/wal/interceptors/segment/manager/partition_manager.go:129]: segment level operation will never generate lots of log. so info level help us to check the seal operation info

File: internal/streamingnode/server/wal/interceptors/segment/manager/partition_manager.go [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+	// Transfer the pending segment into growing state.
+	// Alloc the growing segment at datacoord first.
+	resp, err := resource.Resource().DataCoordClient().AllocSegment(ctx, &datapb.AllocSegmentRequest{
+		CollectionId: pendingSegment.GetCollectionID(),
+		PartitionId:  pendingSegment.GetPartitionID(),
+		SegmentId:    pendingSegment.GetSegmentID(),
+		Vchannel:     pendingSegment.GetVChannel(),
+	})
+	if merr.CheckRPCCall(resp, err); err != nil {
```
[the remark this answers is not in the document]
Comment (chyezh) [reply internal/streamingnode/server/wal/interceptors/segment/manager/partition_manager.go:162]: fix it soon

File: internal/streamingnode/server/wal/interceptors/segment/manager/partition_managers.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+		}
+		if _, ok := metaMaps[rawMeta.GetPartitionId()]; !ok {
+			metaMaps[rawMeta.GetPartitionId()] = make([]*segmentAllocManager, 0, 2)
+		}
+		metaMaps[rawMeta.GetPartitionId()] = append(metaMaps[rawMeta.GetPartitionId()], m)
+	}
+
+	// collectionMap is a map from collectionID to collectionInfo.
+	collectionInfoMap := make(map[int64]*rootcoordpb.CollectionInfoOnPChannel, len(collectionInfos))
+	for _, collectionInfo := range collectionInfos {
```
[the remark this answers is not in the document]
Comment (chyezh) [reply internal/streamingnode/server/wal/interceptors/segment/manager/partition_managers.go:49]: fix it soon

File: internal/streamingnode/server/wal/interceptors/segment/manager/pchannel_manager.go [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+	// recover streaming node growing segment metas.
+	rawMetas, err := resource.Resource().StreamingNodeCatalog().ListSegmentAssignment(ctx, pchannel.Name)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list segment assignment from catalog")
+	}
+	// get collection and parition info from rootcoord.
+	resp, err := resource.Resource().RootCoordClient().GetPChannelInfo(ctx, &rootcoordpb.GetPChannelInfoRequest{
+		Pchannel: pchannel.Name,
+	})
+	if merr.CheckRPCCall(resp, err); err != nil {
```
[the remark this answers is not in the document]
Comment (chyezh) [reply internal/streamingnode/server/wal/interceptors/segment/manager/pchannel_manager.go:36]: will fix it soon

File: internal/streamingnode/server/wal/interceptors/segment/manager/seal_queue.go (hunk 1 above)
Comment (chyezh) [reply internal/streamingnode/server/wal/interceptors/segment/manager/seal_queue.go:177]: same as above, it's a segment level log.

File: internal/streamingnode/server/wal/interceptors/segment/policy/segment_limitation_policy.go (hunk 2 above)
Comment (chyezh) [reply internal/streamingnode/server/wal/interceptors/segment/policy/segment_limitation_policy.go:44]: keep same with datacoord logic

File: internal/streamingnode/server/wal/interceptors/segment/manager/partition_manager.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jaime0815) [reply internal/streamingnode/server/wal/interceptors/segment/manager/partition_manager.go:77]: Minimizing redundant logs is always beneficial.

**Comment by chyezh** [comment]
rerun ut

**Pull request merged**